add BFloat16 support for AdaptiveAvgPool2d on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61015
* #58297
* #56904
* #66927
* #66929
* #56903
* __->__ #56902

Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

mingfeima committed:
add BFloat16 support for AdaptiveAvgPool2d on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836789).

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836789).

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836789).

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836789).

VitalyFedyunin commented on aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp:
```diff
@@ -149,6 +149,105 @@ void cpu_adaptive_avg_pool_channels_last(
 }
 
+template <>
+void cpu_adaptive_avg_pool_channels_last<BFloat16>(
```
Can you please clarify why we need separate kernel code for channels last bfloat16 but can go with same kernel in contiguous

mingfeima replied:
The adaptive {avg|max} pooling kernel is involving reduction.
For `bfloat16` input we should be using `float32` as accumulate type.

On 'contiguous' memory formats, it is purely scalar logic, we don't have a chance to vectorize it so directly casting dtype is proper. But since we can not vectorize the dtype conversion logic, **scalar bfloat16 is going to be slower than original float32** kernel (on most occasions).

On 'channels last' memory formats, we can vectorize the kernel along 'channel' dimension. And for the bfloat16 implementation, we need to allocate a n immediate buffer (float32) to store the 'sum'. Thus the reduction is done on float32 (reduce rounding error and improve performance). I thought mixing bfloat16 kernel with original float32 is going to be a little messy so i wrote a separate <BFLoat16> kernel. And **bfloat16 kernel is faster than float32 when this is a) vectorized kernel; b) memory bandwidth bound kernel**

VitalyFedyunin reviewed (commented):
Please add benchmark numbers

mingfeima commented:
Update with benchmark numbers.

since this patch is not related to parallelism, only single core perf is measured.
this patch has perf change on float32 part, so i only compared the current bfloat16 v.s. float32.

#### avx512 CPU performance
```bash
AdaptiveAvgPool2d(contiguous): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 0.134 ms
AdaptiveAvgPool2d(channels_last): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 0.019 ms
AdaptiveAvgPool2d(contiguous): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 0.182 ms
AdaptiveAvgPool2d(channels_last): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 0.018 ms
AdaptiveAvgPool2d(contiguous): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 15.759 ms
AdaptiveAvgPool2d(channels_last): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 4.354 ms
AdaptiveAvgPool2d(contiguous): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 22.628 ms
AdaptiveAvgPool2d(channels_last): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 2.026 ms
```

Comments:
* on 'contiguous' memory format, bfloat16 is slower than float32 due to it is scalar logic.
* on 'channels last' memory format, bfloat16 uses vectorized logic with smaller dtype conversion overhead. For the smaller size [1, 2048, 7, 7], input is L2 hit (Cascade Lake has L2 of 1MB), so the perf diff is not obvious, while the bigger size would suffer more cache miss, bfloat16 shows perf benefit (1/2 memory size).


#### avx2 CPU performance
```bash
AdaptiveAvgPool2d(contiguous): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 0.186 ms
AdaptiveAvgPool2d(channels_last): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 0.037 ms
AdaptiveAvgPool2d(contiguous): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 0.309 ms
AdaptiveAvgPool2d(channels_last): input_size [1, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 0.040 ms
AdaptiveAvgPool2d(contiguous): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 22.870 ms
AdaptiveAvgPool2d(channels_last): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.float32 time: 6.908 ms
AdaptiveAvgPool2d(contiguous): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 29.968 ms
AdaptiveAvgPool2d(channels_last): input_size [128, 2048, 7, 7] ; output_size:  [2, 2] torch.bfloat16 time: 3.634 ms
```

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima commented:
rebased.

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima commented:
rebased!

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima committed:
Update on "add BFloat16 support for AdaptiveAvgPool2d on CPU"


Differential Revision: [D28836789](https://our.internmc.facebook.com/intern/diff/D28836789)

[ghstack-poisoned]

mingfeima commented:
- [ ] unify bfloat16 code path

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836789).
